### PR TITLE
Revise appveyor CI tests

### PR DESCRIPTION
**Pull request by mikee47**
Further to #2192 this PR updates the appveyor CI scripts:

* Add linux builds
* Split scripts out into setup, build.setup and build.run stages as for travis
* Downloads from espressif are slow, so cache affected files in SmingTools
* Rename `appveyor` to `ci`, and use CI_BUILD_DIR variable to simplify future porting to other CI systems.
* Build documentation in separate stage, add sming-docs.zip artifact

### CODE REVIEW

**Comment by mikee47** [comment]
I was under the impression that the 60 minute time limit applied to the entire run, but the existing builds total 66 minutes so must be per-build. This test approach is perhaps more realistic as we're building in an increasingly 'dirty' VM, which is how development mostly proceeeds.

However, I've experimentally split it back out into separate jobs with a full build of samples, mainly to see how long it takes https://ci.appveyor.com/project/mikee47/sming/builds/37135309.

**Comment by mikee47** [comment]
I've had a brief look at https://www.appveyor.com/docs/build-worker-api/ which potentially could be very useful. I've output messages and test status in this build: https://ci.appveyor.com/project/mikee47/sming/builds/37144433. It does a full samples build for Windows, in total takes almost 3 hours.

**Comment by mikee47** [comment]
@slaff I'm done with this for now I reckon. I've not done anything about deployment stuff though...

**Comment by slaff** [comment]
> I'm done with this for now

@mikee47 Thank you! I will make a new release in the coming days and merge this PR after that.

**Comment by slaff** [comment]
>  will make a new release in the coming days and merge this PR after that.

I was hoping that Travis support will allow us the additional OSS credit that I have requested to at least do the release. But it seems to take time... We have automated a lot of things that are happening for a new release and I would prefer to wait a bit longer instead of having to do all of the things myself...

With that said we should also research possibilities to use Github Actions with our own server (I have one server that is idling and laying around).

**Comment by slaff** [comment]
@mikee47 Can you add also the execution of the deploy.sh script on a new tag ? 

Concerning the secrets: I will try to create a list of all secrets that have to be made available to the CI system. Maybe we can use an encrypted file (openssl aes ) which has per line key=value with the secrets that the CI scripts need. Then we should only take care to provide to the CI system the password/passphrase for decrypting the content of the file.

**Comment by mikee47** [comment]
> 
> @mikee47 Can you add also the execution of the deploy.sh script on a new tag ?
> 
> Concerning the secrets: I will try to create a list of all secrets that have to be made available to the CI system. Maybe we can use an encrypted file (openssl aes ) which has per line key=value with the secrets that the CI scripts need. Then we should only take care to provide to the CI system the password/passphrase for decrypting the content of the file.

@slaff Sure, I'll take a look at that. I'm also reviewing the installation documentation (as per #2099) and re-thinking the installation procedure itself. For Windows, we have Chocolatey but for Linux there are quite a few manual steps especially when ESP32 support is added. The simplest approach may be to just install everything (ESP8266, ESP32, Host, Sming) like this:

```
sudo apt-get install git
git clone https://SmingHub/Sming
Sming/install.sh
```

This would keep the installation instructions minimal. The installation script(s) can also be used for setting up CI rather than duplicating everything.

BUT, a better approach might be to add a Linux repo. to SmingHub and package things up? I've never done any of this so let me know your thoughts?

**Comment by mikee47** [comment]
@slaff Could this be useful http://0install.net/ ?

**Comment by mikee47** [comment]
@slaff OK deployment is working, here's fork run https://ci.appveyor.com/project/mikee47/sming/builds/37303448/job/1qgupdwsspl6oa8h
Just needs updating with appropriate secure keys to update chocolatey repo. and push to github.

**Comment by slaff** [comment]
> OK deployment is working, here's fork run

@mikee47 Awesome! Is this PR ready for merging?

**Comment by slaff** [comment]
> This would keep the installation instructions minimal. The installation script(s) can also be used for setting up CI rather than duplicating everything.

I like the idea. 

> The simplest approach may be to just install everything

I would suggest the default to install everything but to be able by passing parameters to install.sh to be able to choose toolchains to install.

@mikee47 Let's have this in a separate PR once this is merged.

**Comment by mikee47** [comment]
> 
> > OK deployment is working, here's fork run
> 
> @mikee47 Awesome! Is this PR ready for merging?

@slaff Not quite, that's a testing branch. If you can get me the required tokens I'll paste them into this branch and update then should be ready to go.

**Comment by slaff** [comment]
> If you can get me the required tokens I'll paste

@mikee47 Check your email.

**Comment by slaff** [comment]
> Could this be useful http://0install.net/ ?

I had a brief look at it and this might be a very good idea. Our users, especially the Windows ones that are having issues with every new release, can give it a try.

**Comment by mikee47** [comment]
@slaff OK, now working https://ci.appveyor.com/project/mikee47/sming/builds/37340193

Last thing is to encode `SMING_SECRET` in your appveyor account https://ci.appveyor.com/tools/encrypt, then I'll push to this PR.
I've emailed you details.

**Comment by mikee47** [comment]
@slaff I've done a couple of forced-pushes, can I suggest enabling the `Rolling builds` setting in appveyor?

https://www.appveyor.com/docs/build-configuration/#rolling-builds
https://ci.appveyor.com/project/slaff/sming-sb483/settings

**Comment by slaff** [comment]
> can I suggest enabling the Rolling builds setting in appveyor?

Sure, no problem. Should be enabled now.

image

File: Sming/Arch/Host/Tools/ci/build.setup.sh [hunk 1]
```diff
@@ -11,11 +11,13 @@ if [ "$DIFFS" != "" ]; then
 fi
 
 # Make deployment keys, etc. available
-set +x
-openssl aes-256-cbc -d -a -iter 100 -in $CI_BUILD_DIR/.ci/secrets.sh.enc -out /tmp/secrets.sh -pass pass:$SMING_SECRET
-source /tmp/secrets.sh
+if [ -z "$CI_PULL_REQUEST" ]; then
```
Comment (slaff) [inline Sming/Arch/Host/Tools/ci/build.setup.sh:14]: See this link: https://help.appveyor.com/discussions/questions/2615-secure-variables-for-pull-request

Comment (mikee47) [reply Sming/Arch/Host/Tools/ci/build.setup.sh:14]: Yes, the check is necessary though to prevent the decryption failing which kills the build.

**Comment by slaff** [comment]
@mikee47 Do you plan to make more changes to this PR or I shall merge it?

**Comment by mikee47** [comment]
@slaff All done thanks!

**Comment by slaff** [comment]
@mikee47 The coverity scan did not trigger. In the old code if the commit message was having [scan:coverity] or it was having a tag then a coverity scan was triggered: https://ci.appveyor.com/project/slaff/sming-sb483/build/job/gy8k29dr93504ggo?fullLog=true .

**Comment by mikee47** [comment]
@slaff From the log file I see that CI_PULL_REQUEST is undefined, because APPVEYOR_PULL_REQUEST_HEAD_REPO_NAME is also undefined. That additional correction should sort things out.

**Pull request merged**